mavros_extras/src/plugins/odom.cpp

```diff
@@ -64,134 +65,154 @@ class OdometryPlugin : public plugin::PluginBase {
+		/** Pose 6-D Covariance matrix
+		 *  WRT local frame
 		 */
-		Eigen::Matrix<double, 9, 9> r_9d;
-		r_9d.setZero(); // initialize with zeros
-		r_9d.block<3, 3>(0, 0) = r_9d.block<3, 3>(3, 3) = tf_parent2ned.linear();
-		r_9d.block<3, 3>(6, 6) = Eigen::Matrix3d::Identity();
+		Eigen::Matrix<double, 6, 6> r_pose;
+		r_pose.setZero();	// initialize with zeros
+		r_pose.block<3, 3>(0, 0) = r_pose.block<3, 3>(3, 3) = tf_parent2local.linear();
 
-		cov_full_map = r_9d * cov_full_map * r_9d.transpose();
+		// apply the transform
+		cov_pose_map = r_pose * cov_pose_map * r_pose.transpose();
 
-		auto urt_view = Eigen::Matrix<double, 9, 9>(cov_full_map.triangularView<Eigen::Upper>());
-		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: pose+accel covariance matrix: " << std::endl << cov_full_map);
-		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: Cov URT: " << std::endl << urt_view);
+		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: pose covariance matrix:" << std::endl << cov_pose_map);
 
 		/** Velocity 6-D Covariance matrix
 		 *  WRT body frame
 		 */
 		Eigen::Matrix<double, 6, 6> r_vel;
-		r_vel.block<3, 3>(0, 0) = r_vel.block<3, 3>(3, 3) = tf_child2ned.linear();
+		r_vel.setZero();// initialize with zeros
+		r_vel.block<3, 3>(0, 0) = r_vel.block<3, 3>(3, 3) = tf_child2local.linear();
 
+		// apply the transform
 		cov_vel_map = r_vel * cov_vel_map * r_vel.transpose();
 
-		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: velocity covariance matrix: " << std::endl << cov_vel_map);
-
-		// get timestamp
-		uint64_t stamp_usec = odom->header.stamp.toNSec() / 1e3;
+		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: velocity covariance matrix:" << std::endl << cov_vel_map);
 
-		const auto zerov3f = Eigen::Vector3f::Zero();
+		/* -*- ODOMETRY msg parser -*- */
+		mavlink::common::msg::ODOMETRY msg {};
 
-		/* -*- LOCAL_POSITION_NED_COV parser -*- */
-		mavlink::common::msg::LOCAL_POSITION_NED_COV lpos {};
+		msg.time_usec = odom->header.stamp.toNSec() / 1e3;
 
-		lpos.time_usec = stamp_usec;
-		lpos.estimator_type = utils::enum_value(estimator_type);
-		ftf::covariance_urt_to_mavlink(cov_full_map, lpos.covariance);
+		/** Handle frame_id naming - considering the ROS msg frame_id
+		 * as "odom" by default
+		 */
+		if (local_frame_orientation == "ned") {
+			if (frame_source == "vision")
+				msg.frame_id = utils::enum_value(MAV_FRAME::VISION_NED);
+			else if (frame_source == "mocap")
+				msg.frame_id = utils::enum_value(MAV_FRAME::MOCAP_NED);
+			else {
+				ROS_ERROR_NAMED("odom", "Odometry: invalid frame source \"%s\"", frame_source.c_str());
+				return;
+			}
+		}
+		else if (local_frame_orientation == "enu") {
+			if (frame_source == "vision")
+				msg.frame_id = utils::enum_value(MAV_FRAME::VISION_ENU);
+			else if (frame_source == "mocap")
+				msg.frame_id = utils::enum_value(MAV_FRAME::MOCAP_ENU);
+			else {
+				ROS_ERROR_NAMED("odom", "Odometry: invalid frame source \"%s\"", frame_source.c_str());
+				return;
+			}
+		}
+		else {
```

Comment:
I think that may be a bit simplifyed by that method:
```cpp
std::unordered_map<const std::string, MAV_FRAME> m{{
 { "vision-enu", MAV_FRAME:VISION_ENU },
...
};

auto frame_it = m.find(local_frame_orientation + "-" + frame_source);
...
```

Reply:
Yeah sounds good.